core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,430 @@
+            else:
+                exp_commit_log = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (exp_id, version))
+                change_list = exp_commit_log.commit_cmds
+
+                # Handling state additions, renames and deletions.
+                for change_dict in change_list:
+                    if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+                        state_stats_mapping[change_dict['state_name']] = (
+                            stats_domain.StateStats.create_default())
+                    elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                        state_stats_mapping.pop(change_dict['state_name'])
+                    elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                        state_stats_mapping[change_dict['new_state_name']] = (
+                            state_stats_mapping.pop(
+                                change_dict['old_state_name']))
+
+            # Dict mapping each session ID to the list of state hit events for
+            # that session.
+            state_hit_events_session_id_mapping = {}
+            # Dict mapping each state name of the exploration and the list of
+            # unique session IDs that have entered that state.
+            state_session_ids = {
+                state_name: [] for state_name in versioned_exploration.states
+            }
+            session_ids_set = set()
+            # Compute total_hit_count and first_hit_count for the states.
+            for state_hit_event in state_hit_events_for_this_version:
+                state_name = state_hit_event['state_name']
+                state_stats_mapping[state_name].total_hit_count_v1 += 1
+                if state_hit_event['session_id'] not in state_session_ids[
+                        state_name]:
+                    state_stats_mapping[state_name].first_hit_count_v1 += 1
+                    state_session_ids[state_name].append(state_hit_event[
+                        'session_id'])
+                    session_ids_set.add(state_hit_event['session_id'])
+                if state_hit_event['session_id'] not in (
+                        state_hit_events_session_id_mapping):
+                    state_hit_events_session_id_mapping[state_hit_event[
+                        'session_id']] = []
+                state_hit_events_session_id_mapping[state_hit_event[
+                    'session_id']].append(state_hit_event)
+
+            # Compute num_completions for the states.
+            for session_id in session_ids_set:
+                state_hit_event_dicts_sessioned = (
+                    state_hit_events_session_id_mapping[session_id])
+                # All states in the above list except the last element (only if
+                # complete exploration event doesn't exist for the session ID)
+                # are counted as completed because they are ordered by their
+                # timestamp.
```

Comment:
I might be missing something. At which point are you ordering things by timestamp?

Also, the explanation you gave sounds reasonable. I recommend including more of it in the comment here.

Reply:
So, the state_hit_events_dict is first sorted by order of timestamp. Then, we compute state_hit_events_for_this_version, which is also sorted in order of timestamp. Then, for each state hit event in the above list, we add it to a dict keyed by its sessionID (state_hit_events_session_id_mapping). So, for each sessionID in the above dict, all corresponding state hit events are ordered by their timestamp.

Re including more details in the comment, Done.